### DISCUSSION TITLE
issues in modbus library installation

### DISCUSSION THREAD

**Discussion by srinivas1234567899** [pymodbus-dev/pymodbus#1952] [Help] [answered]
Iam not able to download pymodbus library in raspberry pi 4 model b,plz help me in intalling the library

**Comment by janiversen** [comment, marked as answer]
Why are you not able to download it ? You give us no information that enables us to help you, I install regulary on RPI 4 both with Ubuntu and Raspbian without problems.

Please read our guideline in README.

**Reply by srinivas1234567899** [reply]
sir thank you for responding.
This is the code I used.....
from pymodbus.client.sync import ModbusSerialClient as ModbusClient
from pymodbus.constants import Endian
from pymodbus.payload import BinaryPayloadDecoder

UNIT = 0x1

client= ModbusClient(method = "rtu", port="",stopbits = 1, bytesize = 8, parity = 'E', baudrate= 9600)
#modify port with com port of the USB device (eg COM3 in windows or /dev/ttyUSB0 in linux)
client.connect()
response = client.read_holding_registers(3926, 2, unit=UNIT)
decoder = BinaryPayloadDecoder.fromRegisters(response.registers, Endian.Big, wordorder=Endian.Little)
print ("read_holding_registers: " + str(decoder.decode_32bit_float()))
  
client.close()

But when I run the code in thonny it is showing there is no library called pymodbus.client.sync , so I tryed to install pymodbus in terminnal using pip install pymodbus , but it is saying to create a virtual environment or to download it through external source ,  as iam new to work on raspberry pi and modbus iam not able to start it.....

**Reply by janiversen** [reply]
Your code does not correspond to newest version of pymodbus, the code is for an old version of pymodbus. A good way is to read the documentation and use the examples....we have made a few called "simple_*.py" they are there to get you started.

pymodbus.client.sync is NOT a library but a module in the pymodbus library.

If you want to develop with python of course you to setup a python development environment, that is quite normal.

**Reply by srinivas1234567899** [reply]
import time
import minimalmodbus
import serial
import paramiko
import datetime

try:
    client = paramiko.SSHClient()
    client.load_system_host_keys()
    client.set_missing_host_key_policy(paramiko.WarningPolicy())
    client.connect("ip", port, "username", "password")
    dir_name = "/home/sglab/Dropbox/meterdata/new"
    sftp = client.open_sftp()
    sftp.mkdir(dir_name)
except IOError:
    print('Connection with server failed!!!')
    pass

slave_id_of_meters = [101]
instruments = []
files = []

for x in range(len(slave_id_of_meters)):
    instruments.append(minimalmodbus.Instrument('/dev/ttyUSB0', slave_id_of_meters[x]))
    instruments[x].serial.baudrate = 9600
    instruments[x].serial.bytesize = 8
    instruments[x].serial.parity = serial.PARITY_EVEN
    instruments[x].serial.stopbits = 1
    instruments[x].serial.timeout = 1
    instruments[x].mode = minimalmodbus.MODE_RTU
    files.append(sftp.open(dir_name + '/' + str(slave_id_of_meters[x]) + 'test', 'a'))

while True:
    try:
        for x in range(len(slave_id_of_meters)):
            before_read = datetime.datetime.now()
            test_reg = instruments[x].read_registers(0, 2,

This code also I used sir , but Iam not able to download minimalmodbus also sir , the out put is .....no such file detected. 

My aim is to communicate values from multifuntion meter to my raspberry pi using usb to rs485.....Just I want the reading of multimeter as my output in the thonny...

**Reply by janiversen** [reply]
that is another project, please talk to them.

As I wrote please read our documentation at least one of "simple_*" examples does what you want. In respect to downloading pymodbus, just follow the README.

**Reply by srinivas1234567899** [reply]
thank you
